File: apps/api/app/schemas/events.py
"""Event ingestion schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from common.enums import EventStatus


class EventIn(BaseModel):
    customer_id: str = Field(min_length=1, max_length=255, description="Your own customer id")
    event_type: str = Field(min_length=1, max_length=128)
    data: dict[str, Any] = Field(default_factory=dict)
    external_event_id: str | None = Field(default=None, max_length=255)
    occurred_at: datetime | None = None
    customer_email: str | None = Field(default=None, max_length=320)
    customer_name: str | None = Field(default=None, max_length=255)
    source: str = Field(default="api", max_length=64)

    @field_validator("event_type")
    @classmethod
    def _normalize_type(cls, value: str) -> str:
        return value.strip().lower().replace(" ", "_")


class EventPreviewIn(BaseModel):
    """A dry run. Same shape as an event, minus the fields that only matter once stored."""

    customer_id: str = Field(min_length=1, max_length=255, description="Your own customer id")
    event_type: str = Field(min_length=1, max_length=128)
    data: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None

    @field_validator("event_type")
    @classmethod
    def _normalize_type(cls, value: str) -> str:
        return value.strip().lower().replace(" ", "_")


class MemoryPlanOut(BaseModel):
    """One statement the engine found, and what it would do with it."""

    content: str
    type: str
    action: str = Field(description="create, merge, update, supersede, conflict or ignore")
    reason: str
    importance: float = 0.0
    confidence: float = 0.0
    similarity: float = 0.0
    rule: str | None = None
    memory_id: str | None = Field(
        default=None,
        description="The memory this became. Null in a preview, which creates nothing.",
    )
    closest_memory_id: str | None = Field(
        default=None,
        description=(
            "The existing memory this was compared against: the one being changed when "
            "the action is not create, and the nearest miss when it is."
        ),
    )
    closest_content: str | None = None
    sensitivity: str = "normal"
    restricted_by: str | None = None
    extracted_by: str | None = None


class EntityPlanOut(BaseModel):
    name: str
    type: str
    status: str = Field(description="new, existing or touched")
    entity_id: str | None = None


class RedactionOut(BaseModel):
    kind: str
    count: int


class EventExplanationOut(BaseModel):
    """Why an event did, or would, become a memory — or why it would not."""

    would_process: bool = False
    # The field to read first. Null means the pipeline ran to the end.
    stop_reason: str | None = None
    summary: str = ""
    importance: float = 0.0
    threshold: float = 0.0
    text: str | None = Field(
        default=None,
        description="The text the engine read, after redaction. Preview only.",
    )
    text_length: int = 0
    redacted: bool = False
    redactions: list[RedactionOut] = Field(default_factory=list)
    memories: list[MemoryPlanOut] = Field(default_factory=list)
    memory_count: int = 0
    entities: list[EntityPlanOut] = Field(default_factory=list)
    entity_count: int = 0
    duration_ms: float = 0.0


class EventBatchIn(BaseModel):
    events: list[EventIn] = Field(min_length=1, max_length=500)


class EventAccepted(BaseModel):
    event_id: str
    status: Literal["accepted", "duplicate"] = "accepted"
    customer_id: str
    importance: float = 0.0
    queued: bool = True


class EventBatchAccepted(BaseModel):
    accepted: list[EventAccepted]
    duplicates: int = 0


class EventOut(BaseModel):
    id: str
    project_id: str
    customer_id: str
    event_type: str
    external_event_id: str | None = None
    data: dict[str, Any]
    source: str
    importance: float
    status: EventStatus
    occurred_at: datetime
    created_at: datetime
    processed_at: datetime | None = None
    error: str | None = None
    # Why this event did or did not become a memory. Null for events processed before
    # outcomes were recorded, which is different from "nothing happened".
    outcome: EventExplanationOut | None = None

File: sdk/typescript/src/resources/events.ts
/** Event ingestion. */

import type { HttpClient } from "../client.js";
import type { EventExplanation, TrackEventInput, TrackEventResult } from "../types.js";

interface RawTrackResult {
  event_id: string;
  status: "accepted" | "duplicate";
  customer_id: string;
  importance: number;
  queued: boolean;
}

function toPayload(input: TrackEventInput): Record<string, unknown> {
  return {
    customer_id: input.customerId,
    event_type: input.type,
    data: input.data ?? {},
    external_event_id: input.externalEventId,
    occurred_at:
      input.occurredAt instanceof Date ? input.occurredAt.toISOString() : input.occurredAt,
    customer_email: input.customerEmail,
    customer_name: input.customerName,
    source: input.source ?? "sdk",
  };
}

interface RawPlan {
  content: string;
  type: string;
  action: string;
  reason: string;
  importance: number;
  confidence: number;
  similarity: number;
  rule: string | null;
  memory_id: string | null;
  closest_memory_id: string | null;
  closest_content: string | null;
  sensitivity: string;
  restricted_by: string | null;
  extracted_by: string | null;
}

interface RawExplanation {
  would_process: boolean;
  stop_reason: string | null;
  summary: string;
  importance: number;
  threshold: number;
  text: string | null;
  text_length: number;
  redacted: boolean;
  redactions: { kind: string; count: number }[];
  memories: RawPlan[];
  memory_count: number;
  entities: { name: string; type: string; status: string; entity_id: string | null }[];
  entity_count: number;
  duration_ms: number;
}

function fromExplanation(raw: RawExplanation): EventExplanation {
  return {
    wouldProcess: raw.would_process,
    stopReason: raw.stop_reason,
    summary: raw.summary,
    importance: raw.importance,
    threshold: raw.threshold,
    text: raw.text,
    textLength: raw.text_length,
    redacted: raw.redacted,
    redactions: raw.redactions ?? [],
    memories: (raw.memories ?? []).map((plan) => ({
      content: plan.content,
      type: plan.type,
      action: plan.action,
      reason: plan.reason,
      importance: plan.importance,
      confidence: plan.confidence,
      similarity: plan.similarity,
      rule: plan.rule,
      memoryId: plan.memory_id,
      closestMemoryId: plan.closest_memory_id,
      closestContent: plan.closest_content,
      sensitivity: plan.sensitivity,
      restrictedBy: plan.restricted_by,
      extractedBy: plan.extracted_by,
    })),
    memoryCount: raw.memory_count,
    entities: (raw.entities ?? []).map((entity) => ({
      name: entity.name,
      type: entity.type,
      status: entity.status,
      entityId: entity.entity_id,
    })),
    entityCount: raw.entity_count,
    durationMs: raw.duration_ms,
  };
}

function fromResult(raw: RawTrackResult): TrackEventResult {
  return {
    eventId: raw.event_id,
    status: raw.status,
    customerId: raw.customer_id,
    importance: raw.importance,
    queued: raw.queued,
  };
}

export class Events {
  constructor(private readonly http: HttpClient) {}

  /** Send a single event. Returns as soon as the event is durably stored. */
  async track(input: TrackEventInput): Promise<TrackEventResult> {
    const raw = await this.http.request<RawTrackResult>({
      method: "POST",
      path: "/v1/events",
      body: toPayload(input),
      idempotencyKey: input.externalEventId,
    });
    return fromResult(raw);
  }

  /** What this event *would* do, without sending it.
   *
   * The real pipeline, stopped before it writes: nothing is stored and no memory changes.
   * Read `stopReason` first — when it is set, nothing else happened and it says why.
   *
   * ```ts
   * const preview = await memora.events.preview({
   *   customerId: "cus_1",
   *   type: "support_message",
   *   data: { message: "The sync keeps failing." },
   * });
   * if (!preview.wouldProcess) console.log(preview.stopReason);
   * ```
   */
  async preview(
    input: Pick<TrackEventInput, "customerId" | "type" | "data" | "occurredAt">,
  ): Promise<EventExplanation> {
    const raw = await this.http.request<RawExplanation>({
      method: "POST",
      path: "/v1/events/preview",
      body: {
        customer_id: input.customerId,
        event_type: input.type,
        data: input.data ?? {},
        occurred_at:
          input.occurredAt instanceof Date ? input.occurredAt.toISOString() : input.occurredAt,
      },
    });
    return fromExplanation(raw);
  }

  /** Send up to 500 events in one request. */
  async trackBatch(inputs: TrackEventInput[]): Promise<TrackEventResult[]> {
    const raw = await this.http.request<{ accepted: RawTrackResult[]; duplicates: number }>({
      method: "POST",
      path: "/v1/events/batch",
      body: { events: inputs.map(toPayload) },
    });
    return raw.accepted.map(fromResult);
  }

  async list(params: {
    customerId?: string;
    type?: string;
    status?: string;
    limit?: number;
    offset?: number;
  } = {}): Promise<{ data: unknown[]; total: number }> {
    return this.http.request({
      method: "GET",
      path: "/v1/events",
      query: {
        customer_id: params.customerId,
        event_type: params.type,
        status: params.status,
        limit: params.limit,
        offset: params.offset,
      },
    });
  }
}
